otherlibs/fiber/pool.ml

```diff
@@ -20,26 +30,59 @@ let running t k =
   | Open -> k true
   | Closed -> k false
 
-let create () = { mvar = Mvar.create (); status = Open }
+let create () =
+  { tasks = Queue.create (); runner = Awaiting_run; status = Open }
 
 let task t ~f k =
   match t.status with
   | Closed ->
     Code_error.raise "pool is closed. new tasks may not be submitted" []
-  | Open -> Mvar.write t.mvar (Task f) k
-
-let stream t =
-  Stream.In.create (fun () ->
-      let+ next = Mvar.read t.mvar in
-      match next with
-      | Done -> None
-      | Task task -> Some task)
+  | Open -> (
+    Queue.push t.tasks f;
+    match t.runner with
+    | Running | Awaiting_run -> k ()
+    | Awaiting_resume r ->
+      t.runner <- Running;
+      resume r () k)
 
 let stop t k =
   match t.status with
   | Closed -> k ()
-  | Open ->
+  | Open -> (
     t.status <- Closed;
-    Mvar.write t.mvar Done k
+    match t.runner with
+    | Running | Awaiting_run -> k ()
+    | Awaiting_resume r ->
+      t.runner <- Running;
+      resume r () k)
 
-let run t = stream t |> Stream.In.parallel_iter ~f:(fun task -> task ())
+let run t k =
+  match t.runner with
+  | Awaiting_resume _ | Running ->
+    Code_error.raise "Fiber.Pool.run: concurent calls to run aren't allowed" []
+  | Awaiting_run ->
+    t.runner <- Running;
+    (* The number of currently running fibers in the pool. Only when this
+        number reaches zero we may call the final continuation [k] *)
+    let n = ref 1 in
+    let done_fiber () =
+      decr n;
+      if !n = 0 then k () else end_of_fiber
+    in
+    let rec read t =
+      match Queue.pop t.tasks with
+      | None -> on_finish t
+      | Some v ->
+        incr n;
+        fork (fun () -> v () done_fiber) read_delayed
+    and read_delayed () = read t
+    and suspend_k k =
```

Comment:
Why not inline this function? It's small and used only once. And maybe `on_finish` too? I am often getting lost when reading this recursive definition.

Reply:
It's just a general lack of faith in the optimizations of `ocamlopt` :)

Let's take `read_delayed` as an example. My code produces the following `lambda`:

```
on_finish/696
  (function t/702
    (let (*match*/718 = (field 2 t/702))
      (if *match*/718 (apply done_fiber/690 0)
        (apply (field 5 (global Fiber__Core!))
          suspend_k/695 read_delayed/694)))))
```

If we inline it per your suggestion, we'll get:

```
on_finish/695
  (function t/701
    (let (*match*/718 = (field 2 t/701))
      (if *match*/718 (apply done_fiber/690 0)
        (apply (field 5 (global Fiber__Core!))
          suspend_k/694
          (function param/702[int] ;; <- notice the additional closure we allocate
            (apply read/693 t/701)))))))
```

Now I'm not sure how consequential this is given flambda. But flambda is not really widely adopted yet so I still do these minor optimizations by hand.

Reply:
Here's the clambda output. It's a little uglier, but it confirms the additional allocations:

```
;; my version
(fun camlFiber__Pool__on_finish_696 1
    t/702 env/738
  (let (*match*/718 (field 2 t/702))
    (if *match*/718
      (let (clos/739 (field 4 env/738))
        (apply* camlFiber__Pool__done_fiber_690 
          0 clos/739))
      (apply
        (field 5 (read_symbol camlFiber__Core))
        (offset env/738 -3) (offset env/738 -6))))) 

;; read_delayed inlined
(fun camlFiber__Pool__on_finish_696 1
    t/702 env/739
  (let (*match*/719 (field 2 t/702))
    (if *match*/719
      (let (clos/740 (field 4 env/739))
        (apply* camlFiber__Pool__done_fiber_690 
          0 clos/740))
      (apply
        (field 5 (read_symbol camlFiber__Core))
        (offset env/739 -3)
        (closure ;; extra closure
          (fun camlFiber__Pool__fun_741 1
              param/703[int] env/745
            (let
              (arg/746 (field 3 env/745)
                clos/747 (field 2 env/745))
              (apply* camlFiber__Pool__read_693 
                arg/746 clos/747))) 
          (offset env/739 -9) t/702))))) 
```